### ISSUE TITLE
bad results on cityscape datasets

### ISSUE THREAD

**Issue by TengFeiHan0** [WXinlong/SOLO#11]
I have trained SOLO with cityscape datasets, without modifying any settings except for `datasets` part. Here is my result. Obviously, solo performs very badly on the cityscape. Maybe I need to tune some hyper-parameters. Do you have any cues?

```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.075
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.157
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.068
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.004
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.063
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.166
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.096
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.158
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.168
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.008
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.142
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.335
```

**Comment by WXinlong** [comment]
@TengFeiHan0  Thanks for your feedback. The num_grids S should be larger for Cityscapes, e.g.,  try num_grids=[80,72,...] or larger in Decoupled SOLO. Or you can use different num_grids for width and height, as the input ratio of images in Cityscapes is 2:1.  The scale_ranges for multi-level predictions could also be adjusted.  
And can you share your config for our further discussion?

**Comment by TengFeiHan0** [comment]
I'am so sorry about this naive issue. I forgot to change `num_classes`,that's why I got such bad results.  I will follow your instructions and retrain my model, I will share with your about new results if got. thank you for your quick response.

**Comment by WXinlong** [comment]
@TengFeiHan0  No worries. Looking forward to your further results.

**Comment by TengFeiHan0** [comment]
@WXinlong Yesterday, I got a seemingly correct result when training on the cityscape. 
```

Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.196
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.380
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.165
```
As you said, `num_grids `and `scale_ranges` are two important factors, which  influence our final results directly. So, the larger num_grids are, the better output we have. Are there any other factors? 

Obviously, The above result is still not very good, I'm trying to increase num_grids, hope to get a much better performance.

**Comment by WXinlong** [comment]
@TengFeiHan0  Thanks for your update. Well it's much better.  I see that there is still a gap. Can you provide the config so that I can give further suggestions.

**Comment by TengFeiHan0** [comment]
here is my full [log.txt](https://drive.google.com/file/d/1Q-TQ9P4VgeqnnT9wY2yAXcL-Z3cvBIyY/view?usp=sharing)
and I also list some important options as bellow,

```
bbox_head=dict(
        type='SOLOHead',
        num_classes=9,
        in_channels=256,
        stacked_convs=7,
        seg_feat_channels=256,
        strides=[8, 8, 16, 32, 32],
        scale_ranges=((1, 96), (48, 192), (96, 384), (192, 768), (384, 2048)),
        sigma=0.2,
        num_grids=[48, 32, 24, 16,12],
        cate_down_pos=0,
        with_deform=False,
        loss_ins=dict(
            type='DiceLoss',
            use_sigmoid=True,
            loss_weight=3.0),
        loss_cate=dict(
            type='FocalLoss',
            use_sigmoid=True,
            gamma=2.0,
            alpha=0.25,
            loss_weight=1.0),

img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True, with_mask=True),
    dict(
        type='Resize', img_scale=[(2048, 800), (2048, 1024)], keep_ratio=True),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='Pad', size_divisor=32),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels', 'gt_masks']),
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(2048, 1024),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='Pad', size_divisor=32),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img']),
        ])
]

optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0001)
optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
# learning policy
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=1.0 / 3,
    step=[6, ])
checkpoint_config = dict(interval=1)
# yapf:disable
log_config = dict(
    interval=100,
    hooks=[
        dict(type='TextLoggerHook'),
        # dict(type='TensorboardLoggerHook')
    ])
# yapf:enable
# runtime settings
total_epochs = 8
device_ids = range(4)
dist_params = dict(backend='nccl')
log_level = 'INFO'
work_dir = './work_dirs/solo_cityscape'
```
The above log is corresponding to the previous result, more results( more gpus, larger num_grids is comming)

**Comment by WXinlong** [comment]
@TengFeiHan0  I suggest that you could first adjust the num_grids, and modify the scale_range later.  48 is still too small for Cityscapes. For example, 2048 // 48 = 42, which is too large to seperate small pedestrians and cars in a crowd.  You can try Decoupled SOLO and increase num_grid from 48 to at least 80 as said above.  

And please let me know if you get results of 8-gpu training. I'm not familiar with Cityscapes training. From your config, you use 4 gpus, each with one image.  The batch size is decreased from 16 to 4 while the learning rate is the same. I think you should either decrease the learning rate or increase the batch size.

**Comment by TengFeiHan0** [comment]
dear @WXinlong , I'm working on reimplementing the SOLO using `maskrcnn-benchmark`, and got some problems. I hope you could give me some advice. I found that most parts of your code still be used without any big modifications. when dealing with some small errors(for instance,  `maskrcnn-benchmark` uses a `class(SegmentationMask) `to distribute our `mask `image, while `seg_mask` is represented as an image in SOLO) 

In your implementation, you calculate the mass  of the `mask` by calling an existed function. 
`center_h, center_w = ndimage.measurements.center_of_mass(seg_mask)`
How to transform this line under the framework of `maskrcnn-benchmark`?

**Comment by TengFeiHan0** [comment]
@WXinlong  
I forgot to share new results with you, after increasing the  num_grids, I achieved much better performance. I believe that using a better `backbone `, larger  `num_gids` ...  will lead to more higher performance 
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.313
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.550
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.289
```
key part  of my config
```
bbox_head=dict(
        type='DecoupledSOLOHead',
        num_classes=9,
        in_channels=256,
        stacked_convs=7,
        seg_feat_channels=256,
        strides=[8, 8, 16, 32, 32],
        scale_ranges=((1, 96), (48, 192), (96, 384), (192, 768), (384, 2048)),
        sigma=0.2,
        num_grids=[80, 72, 64, 48, 32],

ptimizer = dict(type='SGD', lr=0.005, momentum=0.9, weight_decay=0.0001)
optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
# learning policy
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=1.0 / 3,
    step=[6, ])
checkpoint_config

# runtime settings
total_epochs = 8
device_ids = range(8)
```

**Comment by WXinlong** [comment]
@TengFeiHan0  Great!  Nice to see the performance boost.  Besides the num_grids, the scale_ranges could also be adjusted and might give some improvements.

**Comment by TengFeiHan0** [comment]
I got it,  a lager `scale_ranges` might result in some improvements. Thank you for your help. I will close this issue.

**Comment by jinfagang** [comment]
@TengFeiHan0 How did u change scale_ranges here?

**Comment by jinfagang** [comment]
@WXinlong  Does large num_grids means slower speed? Does the speed largely decreased by applying these grid settings? @TengFeiHan0

**Comment by Areslala** [comment]
@TengFeiHan0  @WXinlong  I just change the scale_ranges like this:
scale_ranges=((1, 144), (72, 288), (144, 576), (288, 1152), (576, 2048))
but the results decrease. Can you give some advice? thansk  a lot.

**Comment by TengFeiHan0** [comment]
@jinfagang I didn't change the scale_range. I think that tuning the num_grids is enough to get a good performance. If you want you to achieve the best performance, probably scale_change should also be tuned.

**Comment by TengFeiHan0** [comment]
@Areslala please keep the original `scale_ranges`, try to tune the `num_girds`, as larger as possible. Please use the decoupled mode to do this experiment.

**Comment by TengFeiHan0** [comment]
@jinfagang as for your second question, larger `num_grids` require the support from more powerful GPUs. So I recommend you use decoupled mode to test large `num_grids`.

**Comment by jinfagang** [comment]
@TengFeiHan0 thanks for your reply. But I want know how much the speed influenced. btw， I tested solov2 with larger num_grids, the result is far worse than yours. I can't reproduce your result

**Comment by jinfagang** [comment]
I think cityscapes can be overfited, this is how train set eval:

![image](https://user-images.githubusercontent.com/21303438/89140456-447a2680-d574-11ea-8c28-e46d93972a84.png)

@TengFeiHan0  Did u trained on trainset and evaled on valset? Or add both train set and test set to train? Which dataset did u eval? As far as I know, eval on test set will lead to a bug in mmdetection older version which is SOLO based.

**Comment by TengFeiHan0** [comment]
I only trained my model on the training dataset and evaluated it on its val dataset. as for the training speed, according to my memory, yes! larger num_girds requires more time and powerful GPUs. @jinfagang by the way, you result seem like very nice.

**Comment by jinfagang** [comment]
@TengFeiHan0 this is the question! this is eval on training set, while eval on val set results seems bad. So I said it is over fited. As for the reason I don't know, It maybe related with warmup iter num when I train on 2 GPUs

**Comment by TengFeiHan0** [comment]
I think two GPUs are not enough, I trained my model on  8 1080ti. @jinfagang

**Comment by jinfagang** [comment]
@TengFeiHan0 Not 4? I don't quite understand the gap between different GPU num since I adjusted lr already

**Comment by TengFeiHan0** [comment]
according to my experience, the more GPUs you have, the larger num_grids you can adjust. @jinfagang maybe it's not true. because these related experiments have been finished several months ago.

**Comment by Sunyingchun** [comment]
Hi @TengFeiHan0, I refer to your parameter settings to achieve below effect on decoupled_sol_light_r50_fpn_8gpu_3x.py config with 5 categories in cityscapes dataset. my GPU is TITAN V.

 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.285
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.509
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.276
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.030
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.278
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.560
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.192
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.344
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.367
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.059
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.381
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.670
+----------+--------+------------+--------+----------+--------+
| category | AP     | category   | AP     | category | AP     |
+----------+--------+------------+--------+----------+--------+
| car      | 40.333 | pedestrian | 21.075 | truck    | 22.700 |
| bus      | 39.636 | rider      | 18.847 | None     | None   |
+----------+--------+------------+--------+----------+--------+

Now I have a few problem：
1. I don't know why each category AP is lower when AP(IOU=0.5)=0.509. How can improve it? 
2.May be it's stupid, but does the each category AP means mask AP or instances segmentation AP?

I will very appreciate for your reply.

**Comment by Sunyingchun** [comment]
Hi   @WXinlong  @TengFeiHan0 , I refer to above settings to achieve below effect on decoupled_sol_light_r50_fpn_8gpu_3x.py config with 5 categories in cityscapes dataset. my GPU is TITAN V.

Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.285
Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=100 ] = 0.509
Average Precision (AP) @[ IoU=0.75 | area= all | maxDets=100 ] = 0.276
Average Precision (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.030
Average Precision (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.278
Average Precision (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.560
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 1 ] = 0.192
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 10 ] = 0.344
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.367
Average Recall (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.059
Average Recall (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.381
Average Recall (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.670
+----------+--------+------------+--------+----------+--------+
| category | AP | category | AP | category | AP |
+----------+--------+------------+--------+----------+--------+
| car | 40.333 | pedestrian | 21.075 | truck | 22.700 |
| bus | 39.636 | rider | 18.847 | None | None |
+----------+--------+------------+--------+----------+--------+

Now I have a few problem：

I don't know why each category AP is lower when AP(IOU=0.5)=0.509. How can improve it?
2.May be it's stupid, but does the each category AP means mask AP or instances segmentation AP?
I will very appreciate for your reply.

**Comment by jinfagang** [comment]
@Sunyingchun  Actually 1 GPU or less than 8 GPU can not replicate the AP as @TengFeiHan0 .

**Comment by EnternalTwinkle** [comment]
> @Sunyingchun Actually 1 GPU or less than 8 GPU can not replicate the AP as @TengFeiHan0 .

请问，方便把COCO格式真值的cityscapes数据json文件发我一份么，我使用上面的设置得到的mAP不到0.2.我的邮箱qtning@163.com，万分感谢

**Comment by cxyjy** [comment]
@jinfagang 
请问，你们的cityscapes数据是如何处理的，我是按照mmdetection2.0版本转成coco格式，但是运行这里的SOLOV2总是会有问题，有些目标就被过滤掉了，最后mask就会变成空的？？

**Comment by pupu-chenyanyan** [comment]
hi, @jinfagang @syc10-09,I use tools/test_ins.py for test, but cannot  evaluate AP for each class . what shouls i do to  evaluate AP for each class? thanks so much

**Comment by blurmemo** [comment]
> 我只在训练数据集上训练了我的模型，并在其 val 数据集上对其进行了评估。至于训练速度，根据我的记忆，是的！更大的num_girds需要更多的时间和强大的 GPU。@jinfagang顺便说一下，你的结果看起来非常好。

你好可以请教一下自定义数据集相关配置修改吗？我现在的AP在我自己数据集上测得很低是1%
